test/end2end_test.go

```diff
@@ -8184,3 +8185,99 @@ func (s) TestGoAwayStreamIDSmallerThanCreatedStreams(t *testing.T) {
 	goAwayWritten.Fire()
 }
+
+type mockBinaryLogger struct {
+	mml *mockMethodLogger
+}
+
+func newMockBinaryLogger() *mockBinaryLogger {
+	return &mockBinaryLogger{
+		mml: &mockMethodLogger{},
+	}
+}
+
+func (mbl *mockBinaryLogger) GetMethodLogger(string) binarylog.MethodLogger {
+	return mbl.mml
+}
+
+type mockMethodLogger struct {
+	events uint64
+}
+
+func (mml *mockMethodLogger) Log(binarylog.LogEntryConfig) {
+	atomic.AddUint64(&mml.events, 1)
+}
+
+// TestGlobalBinaryLoggingOptions tests the binary logging options for client
+// and server side. The test configures a binary logger to be plumbed into every
+// created ClientConn and server. It then makes a unary RPC call, and a
+// streaming RPC call. A certain amount of logging calls should happen as a
+// result of the stream operations on each of these calls.
+func (s) TestGlobalBinaryLoggingOptions(t *testing.T) {
+	csbl := newMockBinaryLogger()
+	ssbl := newMockBinaryLogger()
+
+	internal.AddGlobalDialOptions.(func(opt ...grpc.DialOption))(grpc.WithBinaryLogger(csbl))
+	internal.AddGlobalServerOptions.(func(opt ...grpc.ServerOption))(grpc.BinaryLogger(ssbl))
+	defer func() {
+		internal.ClearGlobalDialOptions()
+		internal.ClearGlobalServerOptions()
+	}()
+	ss := &stubserver.StubServer{
+		UnaryCallF: func(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+			return &testpb.SimpleResponse{}, nil
+		},
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for {
+				_, err := stream.Recv()
+				if err == io.EOF {
+					return nil
+				}
+			}
+		},
+	}
+
+	// No client or server options specified, because should pick up configured
+	// global options.
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Make a Unary RPC. This should cause Log calls on the MethodLogger.
+	if _, err := ss.Client.UnaryCall(ctx, &testpb.SimpleRequest{}); err != nil {
+		t.Fatalf("Unexpected error from UnaryCall: %v", err)
+	}
+	if csbl.mml.events != 5 {
+		t.Fatalf("want 5 client side binary logging events, got %v", csbl.mml.events)
+	}
+	if ssbl.mml.events != 5 {
+		t.Fatalf("want 5 server side binary logging events, got %v", ssbl.mml.events)
+	}
+
+	// Make a streaming RPC. This should cause Log calls on the MethodLogger.
+	stream, err := ss.Client.FullDuplexCall(ctx)
+	if err != nil {
+		t.Fatalf("ss.Client.FullDuplexCall failed: %f", err)
+	}
+	rpcDone := make(chan struct{})
+	go func() {
+		for {
+			_, err := stream.Recv()
+			if err == io.EOF {
+				close(rpcDone)
+				return
+			}
+		}
+	}()
+	stream.CloseSend()
+	<-rpcDone
+	if csbl.mml.events != 9 {
+		t.Fatalf("want 9 client side binary logging events, got %v", csbl.mml.events)
+	}
+	if ssbl.mml.events != 8 {
+		t.Fatalf("want 8 server side binary logging events, got %v", ssbl.mml.events)
+	}
```

Comment:
Why aren't these the same?

Reply:
This is WAI. I looked over my diff (and found some correctness issues - which you can see in the diff of the commit I'm about to push out). However, my diff from master really keeps the exact same number of logging calls in each operation/over the stream lifetime.

Reply:
Can you explain the difference, though?  It doesn't block this PR, but we need to understand this and make sure it's correct.

Reply:
Yeah, I'm testing this exact scenario in o11y e2e, and writing out the exact atoms of events emitted that we expect like we discussed, hopefully if there's any glaring correctness issues it'll show up then, or we can see it's WAI, and we can look into it further.

Reply:
We can also print them here really easily, right?  Why not just add a couple `t.Log`s and let's see what it is.